File: src/main/java/com/wts/kayan/app/utility/PrimaryUtilities.java
package com.wts.kayan.app.utility;

import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.functions;
import org.apache.spark.sql.types.StructType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;

/**
 * Static utility methods shared across the Spark application.
 *
 * <p><strong>Enhancement vs. previous units:</strong> paths are no longer hardcoded.
 * Every method that previously computed a path from string literals now accepts an
 * {@link AppConfig} instance and reads the path from {@code application.conf}.
 * This makes the same code work in dev, test, and prod without recompilation.
 *
 * @author Mehdi TAJMOUATI
 * @see <a href="https://www.wytasoft.com/wytasoft-group/">WyTaSoft — courses and training sessions</a>
 */
public final class PrimaryUtilities {

    private static final Logger log = LoggerFactory.getLogger(PrimaryUtilities.class);

    private PrimaryUtilities() {}

    // -------------------------------------------------------------------------
    // Partition detection
    // -------------------------------------------------------------------------

    /**
     * Scans a directory and returns the value of the most recent date partition.
     *
     * <p>Looks for sub-directories whose name starts with {@code <columnPartitioned>=}
     * (e.g. {@code date=2023-09-20}), extracts the date portion, and returns the
     * maximum as a {@code yyyy-MM-dd} string.
     *
     * @param path              Directory to scan (local path or HDFS URI).
     * @param columnPartitioned Partition column name (e.g. {@code "date"}).
     * @param spark             Active SparkSession (used to access Hadoop configuration).
     * @return Most recent partition value as {@code yyyy-MM-dd}, or {@code "2999-01-01"}
     *         if no partitions are found.
     */
    public static String getMaxPartition(String path, String columnPartitioned, SparkSession spark) {
        try {
            FileSystem fs = FileSystem.get(spark.sparkContext().hadoopConfiguration());
            FileStatus[] statuses = fs.listStatus(new Path(path));

            String prefix = columnPartitioned + "=";
            String[] partitionValues = Arrays.stream(statuses)
                    .filter(FileStatus::isDirectory)
                    .map(s -> s.getPath().getName())
                    .filter(name -> name.startsWith(prefix))
                    .map(name -> name.substring(prefix.length()))
                    .toArray(String[]::new);

            if (partitionValues.length == 0) {
                log.info("\n**** No partitions by {} found in {} ****\n", columnPartitioned, path);
                return "2999-01-01";
            }

            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
            Date maxDate = Arrays.stream(partitionValues)
                    .map(d -> {
                        try { return sdf.parse(d); }
                        catch (ParseException e) { throw new RuntimeException(e); }
                    })
                    .max(Date::compareTo)
                    .orElseThrow(() -> new RuntimeException("Unexpected empty partition list"));

            String maxPartition = sdf.format(maxDate);
            log.info("\n**** max {} = {} ****\n", columnPartitioned, maxPartition);
            return maxPartition;

        } catch (IOException e) {
            log.error("Fatal Exception: could not read partition data from {}", path);
            throw new RuntimeException(e);
        }
    }

    // -------------------------------------------------------------------------
    // DataFrame reading  (paths now come from AppConfig)
    // -------------------------------------------------------------------------

    /**
     * Reads a CSV dataset into a typed {@link Dataset}.
     *
     * <p>Paths are resolved from {@link AppConfig} instead of being hardcoded,
     * so the same method works across dev / test / prod environments.
     *
     * @param sourceName      Dataset identifier ({@link PrimaryConstants#CLIENTS} or
     *                        {@link PrimaryConstants#ORDERS}).
     * @param schema          Explicit schema — avoids Spark's inferSchema scan.
     * @param isLastPartition When {@code true} only the most recent partition of orders
     *                        is loaded and a {@code date} literal column is appended.
     * @param spark           Active SparkSession.
     * @param appConfig       Resolved application configuration.
     * @return Typed {@link Dataset} loaded from the resolved path.
     */
    public static Dataset<Row> readDataFrame(
            String sourceName,
            StructType schema,
            boolean isLastPartition,
            SparkSession spark,
            AppConfig appConfig) {

        log.info("\n**** Reading DataFrame for {} ****\n", sourceName);

        String inputPath;
        String tableName;
        String partitionValue = "";

        // Paths come from AppConfig — no hardcoded strings.
        switch (sourceName.toLowerCase()) {

            case PrimaryConstants.CLIENTS:
                inputPath = appConfig.getClientsPath();
                tableName = "";
                break;

            case PrimaryConstants.ORDERS:
                inputPath = appConfig.getOrdersPath();
                // Detect the most recent partition folder at runtime.
                partitionValue = getMaxPartition(inputPath, appConfig.getPartitionColumn(), spark);
                tableName      = "date=" + partitionValue + "/";
                break;

            default:
                throw new IllegalArgumentException("Unknown source: " + sourceName);
        }

        String fullPath = inputPath + tableName;
        log.info("\n Loading {} from {} \n", sourceName, fullPath);

        String[] columns = ColumnSelector.getColumnSequence(sourceName);

        Dataset<Row> dataFrame = spark.read()
                .schema(schema)
                .option("header", "true")
                .option("delimiter", ",")
                .csv(fullPath)
                .selectExpr(columns);

        if (isLastPartition && !partitionValue.isEmpty()) {
            log.info("\n Adding partition column '{}' = {} \n",
                    appConfig.getPartitionColumn(), partitionValue);
            return dataFrame.withColumn(appConfig.getPartitionColumn(),
                    functions.lit(partitionValue));
        }

        return dataFrame;
    }

    /**
     * Convenience overload — reads the full dataset without partition filtering.
     */
    public static Dataset<Row> readDataFrame(
            String sourceName,
            StructType schema,
            SparkSession spark,
            AppConfig appConfig) {
        return readDataFrame(sourceName, schema, false, spark, appConfig);
    }

    // -------------------------------------------------------------------------
    // DataFrame writing  (output path now comes from AppConfig)
    // -------------------------------------------------------------------------

    /**
     * Writes a {@link Dataset} to Parquet, partitioned by the column declared in
     * {@code AppConfig} ({@code writer.partition_by}).
     *
     * @param dataFrame  Dataset to persist.
     * @param appConfig  Resolved application configuration.
     */
    public static void writeDataFrame(Dataset<Row> dataFrame, AppConfig appConfig) {

        String mode         = appConfig.getWriterMode();
        int    numPartition = appConfig.getNumPartitions();
        String outputPath   = appConfig.getOutputPath();
        String partitionBy  = appConfig.getPartitionBy();

        log.info("\n *** Write started (mode={}, partitions={}, path={}) ***\n",
                mode, numPartition, outputPath);

        dataFrame
                .coalesce(numPartition)
                .write()
                .format("parquet")
                .partitionBy(partitionBy)
                .mode(mode)
                .save(outputPath);

        log.info("\n *** Write completed *** \n");
    }
}

File: src/main/scala/com/wts/kayan/app/utility/PrimaryUtilities.scala
package com.wts.kayan.app.utility

import com.typesafe.config.Config
import org.apache.spark.sql.{Column, DataFrame, SparkSession}
import org.slf4j.LoggerFactory
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.spark.SparkContext
import org.apache.spark.sql.functions.lit
import org.apache.spark.sql.types.StructType

import java.io.{BufferedReader, InputStreamReader, Reader}
import java.text.SimpleDateFormat
import java.util.Date

/**
 * Utility object providing methods for interacting with HDFS, reading and writing DataFrames,
 * and handling date conversions, tailored for Spark applications. These utilities facilitate
 * common tasks such as reading data, finding the most recent data partitions, and writing data,
 * thereby simplifying data management tasks.
 *
 * @note Use these utilities to enhance code reusability and maintain clean, efficient operations within Spark jobs.
 * @see <a href="https://www.wytasoft.com/wytasoft-group/">Visit WyTaSoft for more information on Spark applications and data processing.</a>
 * @see <a href="https://www.linkedin.com/in/mtajmouati">Mehdi TAJMOUATI's LinkedIn profile</a>
 */
object PrimaryUtilities {

  private val log = LoggerFactory.getLogger(this.getClass)

  /**
   * Fetches the most recent partition based on a date column from a specified HDFS path.
   * This method is useful for incremental data loading scenarios.
   *
   * @param path The HDFS directory to scan.
   * @param columnPartitioned The partition column, defaulted to 'date'.
   * @param spark The Spark session.
   * @return The most recent partition date as a string, or a far-future date if no partitions exist.
   * @throws ArrayIndexOutOfBoundsException if there is an error reading the partition data.
   */
  def getMaxPartition(path: String, columnPartitioned: String = "date")(
    spark: SparkSession): String = {
    val fs = FileSystem.get(spark.sparkContext.hadoopConfiguration)

    try {
      val listOfInsertDates: Array[String] = fs
        .listStatus(new Path(s"$path"))
        .filter(_.isDirectory)
        .map(_.getPath)
        .map(_.toString)
      val filterPartitionFolders =
        listOfInsertDates.filter(_.contains(s"$columnPartitioned="))
      val insertDateStr = filterPartitionFolders.map(_.split(s"$columnPartitioned=")(1))

      if (insertDateStr.length > 1) {
        val dateFormatted =
          insertDateStr.map(d => convertStringToDate(d, "yyyy-MM-dd"))
        val maxDate = dateToStrNdReformat(dateFormatted.max, "yyyy-MM-dd")
        log.info(s"\n**** max $columnPartitioned $maxDate ****\n")
        maxDate
      } else if (insertDateStr.length == 0) {
        log.info(s"\n**** there are no partitions by $columnPartitioned in $path ****\n")
        "2999-01-01"
      } else {
        log.info(s"\n**** max $columnPartitioned ${insertDateStr(0)} ****\n")
        insertDateStr(0)
      }
    } catch {
      case _: Throwable =>
        log.error("Fatal Exception: Check Src-View Data")
        throw new ArrayIndexOutOfBoundsException
    }
  }

  /**
   * Reads a DataFrame from a specified source path using a predefined schema, supporting data partitioning.
   *
   * @param sourceName The identifier for the data source to load.
   * @param schema The schema to apply to the DataFrame.
   * @param isCondition Boolean indicating if a condition should be applied.
   * @param condition The conditional filter to apply, if any.
   * @param sparkSession Implicit SparkSession to handle DataFrame operations.
   * @param env Implicit environment used for building the data path.
   * @param config Implicit Config object for additional settings.
   * @return DataFrame loaded from the specified path.
   */
  def readDataFrame(sourceName: String,
                    schema: StructType,
                    isCondition: Boolean = false,
                    condition: Column = null,
                    isLastPartition: Boolean = false)
                   (implicit sparkSession: SparkSession, env: String, config: Config): DataFrame = {

    log.info(s"\n**** Reading file to create DataFrame ****\n")

    // Generate the effective condition
    val effectiveCondition: Column = if (isCondition) condition else lit(true)

    // Resolve the base input directory from configuration. The config key segment is the
    // sourceName parameter: "clients" -> training_spark.inputs.clients.<env>,
    // "orders" -> training_spark.inputs.orders.<env>.
    val basePath: String = config.getString(s"training_spark.inputs.$sourceName.$env")

    // For orders, read only the most recent date= partition found under the base path.
    var partitionedValue = ""
    val readPath: String = sourceName match {
      case PrimaryConstants.ORDERS =>
        partitionedValue = getMaxPartition(basePath)(sparkSession)
        s"${basePath}date=$partitionedValue/"
      case _ => basePath
    }

    log.info(s"\n Loading $sourceName ($env) from $readPath ***\n")

    val dataFrame: DataFrame = sparkSession.read
      .schema(schema)
      .option("header", "true")       // treat first line as header
      .option("delimiter", ",")       // field separator
      .csv(readPath)
      .selectExpr(ColumnSelector.getColumnSequence(sourceName): _*)
      .where(effectiveCondition)

    if (isLastPartition) {
      log.info(s"\n Loading with partition ***\n")
      return dataFrame.withColumn("date", lit(partitionedValue))
    }

    dataFrame
  }

  /**
   * Opens a file in HDFS and returns a BufferedReader to read the file's contents.
   *
   * @param filePath The full path to the file in HDFS.
   * @param sc The SparkContext to access Hadoop configurations.
   * @return A BufferedReader that can be used to read the file.
   */
  def getHdfsReader(filePath: String)(sc: SparkContext): Reader = {
    val fs = FileSystem.get(sc.hadoopConfiguration)
    val path = new Path(filePath)
    new BufferedReader(new InputStreamReader(fs.open(path)))
  }

  /**
   * Converts a string to a Date object using the specified date format.
   *
   * @param s The date string to convert.
   * @param formatType The format of the date string.
   * @return A Date object representing the parsed date string.
   */
  def convertStringToDate(s: String, formatType: String): Date = {
    val format = new SimpleDateFormat(formatType)
    format.parse(s)
  }

  /**
   * Reformats a Date object to a string using the specified date format.
   *
   * @param date The Date object to reformat.
   * @param format The desired format of the date string.
   * @return A string representing the formatted date.
   */
  def dateToStrNdReformat(date: Date, format: String): String = {
    val df = new SimpleDateFormat(format)
    df.format(date)
  }

  /**
   * Writes a DataFrame to storage using a path resolved from configuration.
   *
   * The output path is read from `training_spark.outputs.<datasetName>.<env>`, keeping the
   * destination configurable per dataset and per environment rather than hard-coded.
   *
   * @param dataFrame The DataFrame to write.
   * @param datasetName The output dataset identifier (e.g. `PrimaryConstants.CLIENTS_ORDERS`),
   *                    used together with `env` to resolve the destination path from config.
   * @param mode The save mode (e.g., "overwrite", "append").
   * @param numPartition The number of partitions to use when writing the DataFrame.
   * @param env Implicit environment string used to resolve the output path.
   * @param config Implicit Config providing the configured output paths.
   */
  def writeDataFrame(dataFrame: DataFrame,
                     datasetName: String,
                     mode: String,
                     numPartition: Int)(implicit env: String, config: Config): Unit = {

    val outputPath = config.getString(s"training_spark.outputs.$datasetName.$env")

    log.info(s"\n *** Write started (dataset: $datasetName, env: $env, path: $outputPath, mode: $mode, numPartition: $numPartition) ... ***\n")

    dataFrame
      .coalesce(numPartition)
      .write
      .format("parquet")
      .partitionBy("location")
      .mode(mode)
      .save(outputPath)

    log.info(s"\n *** Write Completed ... *** \n")
  }
}
